server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -409,28 +440,89 @@ private HealthIndicatorResult getMasterIsStableResult(boolean explain, MasterHis
      * @return The HealthIndicatorResult for the given localMasterHistory
      */
     private HealthIndicatorResult calculateOnHaveNotSeenMasterRecently(MasterHistory localMasterHistory, boolean explain) {
-        // NOTE: The logic in this method will be implemented in a future PR
-        String summary = "No master has been observed recently";
-        Map<String, Object> details = new HashMap<>();
-        List<UserAction> userActions = getContactSupportUserActions(explain);
-        return createIndicator(
-            HealthStatus.RED,
-            summary,
-            explain ? new SimpleHealthIndicatorDetails(details) : HealthIndicatorDetails.EMPTY,
-            UNSTABLE_MASTER_IMPACTS,
-            userActions
+        Collection<DiscoveryNode> masterEligibleNodes = getMasterEligibleNodes();
+        final HealthIndicatorResult result;
+        boolean leaderHasBeenElected = coordinator.getPeerFinder().getLeader().isPresent();
+        if (masterEligibleNodes.isEmpty() && leaderHasBeenElected == false) {
+            result = calculateOnNoMasterEligibleNodes(localMasterHistory, explain);
+        } else if (leaderHasBeenElected) {
+            DiscoveryNode currentMaster = coordinator.getPeerFinder().getLeader().get();
+            result = calculateOnCannotJoinLeader(localMasterHistory, currentMaster, explain);
+        } else {
+            // NOTE: The logic in this block will be implemented in a future PR
+            result = createIndicator(
+                HealthStatus.RED,
+                "No master has been observed recently",
+                HealthIndicatorDetails.EMPTY,
+                UNSTABLE_MASTER_IMPACTS,
+                getContactSupportUserActions(explain)
+            );
+        }
+        return result;
+    }
+
+    /**
+     * Creates a HealthIndicatorResult in the case that there has been no master in the last few seconds, there is no elected master known,
+     * and there are no master eligible nodes. The status will be RED, and the details (if explain is true) will contain the list of any
+     * masters seen previously and a description of known problems from this node's Coordinator.
+     * @param localMasterHistory Used to pull recent master nodes for the details if explain is true
+     * @param explain If true, details are returned
+     * @return A HealthIndicatorResult with a RED status
+     */
+    private HealthIndicatorResult calculateOnNoMasterEligibleNodes(MasterHistory localMasterHistory, boolean explain) {
```

Comment:
nit: This `calculate` method is a bit different than the rest in this service as it's just creating the indicator result. Should we name it something to reflect that? ie. `getIndicatorResultOnNoMasterEligibleNodes`

Same for `calculateOnCannotJoinLeader`